bin/propolis-server/src/lib/spec/config_toml.rs

```diff
@@ -0,0 +1,277 @@
+    name: &str,
+    backend: &config::BlockDevice,
+) -> Result<StorageBackendV0, ConfigTomlError> {
+    let backend_spec = match backend.bdtype.as_str() {
+        "file" => StorageBackendV0::File(FileStorageBackend {
+            path: backend
+                .options
+                .get("path")
+                .ok_or_else(|| {
+                    ConfigTomlError::InvalidFileBackendPath(name.to_owned())
+                })?
+                .as_str()
+                .ok_or_else(|| {
+                    ConfigTomlError::InvalidFileBackendPath(name.to_owned())
+                })?
+                .to_string(),
+            readonly: match backend.options.get("readonly") {
+                Some(toml::Value::Boolean(ro)) => Some(*ro),
+                Some(toml::Value::String(v)) => {
+                    Some(v.parse::<bool>().map_err(|e| {
+                        ConfigTomlError::FileBackendReadonlyParseFailed(
+                            name.to_owned(),
+                            e,
+                        )
+                    })?)
+                }
+                _ => None,
+            }
+            .unwrap_or(false),
+        }),
+        _ => {
+            return Err(ConfigTomlError::InvalidStorageBackendType(
+                backend.bdtype.clone(),
+                name.to_owned(),
+            ));
+        }
+    };
+
+    Ok(backend_spec)
+}
+
+pub(super) fn parse_storage_device_from_config(
+    name: &str,
+    device: &config::Device,
+) -> Result<StorageDeviceV0, ConfigTomlError> {
+    enum Interface {
+        Virtio,
+        Nvme,
+    }
+
+    let interface = match device.driver.as_str() {
+        "pci-virtio-block" => Interface::Virtio,
+        "pci-nvme" => Interface::Nvme,
+        _ => {
+            return Err(ConfigTomlError::InvalidStorageDeviceType(
+                device.driver.clone(),
+                name.to_owned(),
+            ));
+        }
+    };
+
+    let backend_name = device
+        .options
+        .get("block_dev")
+        .ok_or_else(|| {
+            ConfigTomlError::NoBackendNameForStorageDevice(name.to_owned())
+        })?
+        .as_str()
+        .ok_or_else(|| {
+            ConfigTomlError::NoBackendNameForStorageDevice(name.to_owned())
+        })?
+        .to_owned();
+
+    let pci_path: PciPath = device
+        .get("pci-path")
+        .ok_or_else(|| ConfigTomlError::InvalidPciPath(name.to_owned()))?;
+
+    Ok(match interface {
+        Interface::Virtio => {
+            StorageDeviceV0::VirtioDisk(VirtioDisk { backend_name, pci_path })
+        }
+        Interface::Nvme => {
+            StorageDeviceV0::NvmeDisk(NvmeDisk { backend_name, pci_path })
+        }
+    })
+}
+
+pub(super) struct ParsedNetworkDevice {
+    pub(super) device_name: String,
+    pub(super) device_spec: NetworkDeviceV0,
+    pub(super) backend_name: String,
+    pub(super) backend_spec: NetworkBackendV0,
+}
+
+pub(super) fn parse_network_device_from_config(
```

Comment:
I'm not sure I follow--in this case I think we need more than just a `&str` anyway, since we've also got the previously-parsed device information from the TOML to contend with. (Did you have a different type in mind? We could maybe do something like that for the network backend type, although this code is going to slosh around a fair bit more before it's done, so I'd probably do that in a later PR.)

Reply:
whoops, i can't read. disregard this one.